ref-arch/real-time-analytics.rst

```diff
@@ -0,0 +1,509 @@
+  );
+
+Since this function is given some metadata from the master, where does the master get that
+metadata from? Every minute it calls its own function which fires off all the
+aggregations:
+
+.. code-block:: plpgsql
+
+  -- this should be run on the master
+  CREATE FUNCTION run_rollups(source_table text, dest_table text) RETURNS void
+  AS $$
+  DECLARE
+  BEGIN
+    FOR source_shard, dest_shard, nodename, nodeport IN
+      SELECT
+        a.logicalrelid::regclass||'_'||a.shardid,
+        b.logicalrelid::regclass||'_'||b.shardid,
+        nodename, nodeport
+      FROM pg_dist_shard a
+      JOIN pg_dist_shard b USING (shardminvalue)
+      JOIN pg_dist_shard_placement p ON (a.shardid = p.shardid)
+      WHERE a.logicalrelid = 'first'::regclass AND b.logicalrelid = 'second'::regclass;
+    LOOP
+      SELECT * FROM dblink(
+        format('host=%s port=%d', nodename, nodeport),
+        format('SELECT rollup_1min(%, %s);', source_shard, dest_shard));
+    END LOOP;
+  END;
+  $$ LANGUAGE 'plpgsql';
+
+.. NOTE::
+
+  There are many ways to make sure the function is called periodically and no answer that
+  works well for every system. If you're able to run cron on the same machine as the
+  master, you can do something as simple as this:
+
+  .. code-block:: bash
+  
+    * * * * * psql -c "SELECT run_rollups('http_requests', 'http_requests_1min');"
+
+The dashboard query from earlier is now a lot nicer:
+
+.. code-block:: sql
+
+  SELECT
+    request_count, success_count, error_count, average_response_time_msec
+  FROM http_request_1min
+  WHERE zone_id = 1 AND minute = date_trunc('minute', now());
+
+Expiring Old Data
+-----------------
+
+The rollups make queries faster but we still have a lot of raw data sitting around. How
+long you should keep each granularity of data is a business decision, but once you decide
+it's easy to write a function to expire old data:
+
+.. code-block:: plpgsql
+
+  -- another function for the master
+  CREATE FUNCTION expire_old_request_data RETURNS void
+  AS $$
+    SET citus.all_modification_commutative TO TRUE;
+    SELECT master_modify_multiple_shards(
+      'DELETE FROM http_request WHERE ingest_time < now() - interval ''1 hour'';');
+    SELECT master_modify_multiple_shards(
+      'DELETE FROM http_request_1min WHERE ingest_time < now() - interval ''1 day'';');
+    RESET citus.all_modification_commutative;
+  END;
+  $$ LANGUAGE 'sql';
+
+.. NOTE::
+
+  The above function should be called every minute. As mentioned above there are many
+  different ways to accomplish this and no way which makes everybody happy. If you're
+  capable of adding cron entries to the machine the master is running on you might
+  consider adding a crontab entry:
+
+  .. code-block:: bash
+  
+    * * * * * psql -c "SELECT expire_old_request_data();"
+
+Review, what have we done?
+--------------------------
+
+That's the entire architecture! The next few sections are just extensions for additional
```

Comment:
The word extensions is a bit confusing given the PostgreSQL context. Maybe just: The next few sections provide solutions for additional problems that often pop up.